File: layers/combinelayer.py
from utils.consts    import *
from utils.fmapdesc  import OfmapDesc
from layer           import Layer
import nets.network

##########################################################
class CombineLayer(Layer): # abstract class
    #-----------------------------------------------------------------
    def __init__(self, layerName, ntwk, prev_layer, earlier_layer, num_ofmaps):
        assert(isinstance(ntwk, nets.network.Network))
        assert(isinstance(prev_layer, Layer))
        assert(isinstance(earlier_layer, Layer))
        assert(prev_layer.gOfmapSize() == earlier_layer.gOfmapSize())

        ofmap_desc = OfmapDesc(num_ofmaps, prev_layer.gOfmapSize())

        super(CombineLayer, self).__init__(layerName, ntwk, (prev_layer, earlier_layer), ofmap_desc)

    #-----------------------------------------------------------------
    def qPassThrough(self):
        return False

File: compiler/be/layers/src/combinelayer.cpp
#include "utils/inc/asserter.hpp"
#include "layers/inc/combinelayer.hpp"


namespace kcc {

namespace nets {
    class Network;
}

namespace layers {

//--------------------------------------------------------
CombineLayer::CombineLayer(const Params& params, Layer* prev_layer, Layer* earlier_layer, kcc_int32 num_ofmaps)
    : Layer(params,
            FmapDesc(num_ofmaps, prev_layer->gOfmapHeight(), prev_layer->gOfmapWidth()),
            mkLayerVector2(prev_layer, earlier_layer))

{
    Assert(num_ofmaps > 0, "Combine layer: number of OFMAPs must be positive");

    Assert(prev_layer->gOfmapWidth() == earlier_layer->gOfmapWidth(),
           "Combine layer: Input image widths not identical");
    Assert(prev_layer->gOfmapHeight() == earlier_layer->gOfmapHeight(),
           "Combine layer: Input image heights not identical");
}

//----------------------------------------------------------------
bool
CombineLayer::qPassThrough() const
{
    return false;
}

//----------------------------------------------------------------
bool
CombineLayer::qCombineLayer() const
{
    return true;
}

} // namespace layers
} // namespace kcc

